### PR TITLE
TAP-based GI tests

### PR DESCRIPTION
**Pull request by LordYuuma**
This patch replaces the large everything folder by two test files, which generate roughly the same results, but through TAP. Within these files, the test patterns are much more obvious than before. Eventually, we should be able to generate them from the GIR through pattern matching (as implied by #36).

### CODE REVIEW

**Comment by spk121** [comment]
This is certainly shorter and cleaner.

I just recently noticed that automake supplies a test harness for scheme files, based on srfi-64, which apparently originally comes from Guix.
https://git.savannah.gnu.org/cgit/automake.git/tree/contrib/test-driver.scm

It is used in Guile-GCrypt, for example
https://notabug.org/cwebber/guile-gcrypt

If we are going to compress the tests down, do you have an opinion about TAP-based over srfi-64 based tests?  One advantage to TAP is that it prints a line for each test during `make check`, while srfi-64 tests with this test harness print only one result for a compete file, leaving you to investigate the test-suite.log.  The advantage to srfi-64 would be that some other scheme programs use it.

**Comment by LordYuuma** [comment]
To be honest, my bindings are quick and dirty – something based on SRFI-64 will have a conceptual advantage over that. However, as far as I can see, there are some problems.
1. One result per file. This is exactly what I sought to solve with my custom TAP bindings. However, that does not mean we can't write a test-driver, that has this based on SRFI-64, only that this specific driver likely is not what we're looking for. (Although I'm pretty sure, that `test-on-test-end-gnu` should report one result per test instead of per file, perhaps that's a bug?)
2. Missing "ERROR". Our tests are normally wrapped in a `hard-error-if-exception` and also produce an ERROR if we abort or segfault on C. Such conditions can – at least according to this driver – not be replicated by SRFI-64. This is not too big of a problem though, especially if we can accept exceptions as soft errors.
3. How to environment. Since the driver loads the script, we'd have to set up the environment around the driver, not the script. I'm pretty sure, guile-gcrypt also has the same issue and we can just copy their solution to it, though.

In summary, SRFI-64 would definitely be cleaner than this, but also requires a bit more setup. However, this patchset already conflicts with your work on `dev-marshalling-tests-3` to an extent (since you add everything tests while I remove them), so at least some rewriting must inevitably take place. If you want to merge those first – given that they solve #54, that might be a good idea – you might even go to SRFI-64 directly. However, migrating from the bindings of this patch to SRFI-64 should not be that hard either. In essence, we'd have to rewrite `(tap:test! name body ...)` to `(test-assert name (begin body ...))` whenever body covers multiple expressions.

**Comment by spk121** [comment]
In the near term, the one-test-per-script method has two advantages, for me.
- If I need to run gdb on a script to check the behaviour of a particular C function to a particular test, it is simpler to run.
- If I use `tools/valgrind.sh` to check memory management, the smaller test case helps isolate problems.
Both of those reasons go away when things are stable, but, they are still important to me right now.

**Comment by LordYuuma** [comment]
I think this need would be better served by test script generation à la #36. If we wrote a generator script with filtering capabilities, we could create both TAP/SRFI-64-based tests and test suites in the same manner.

With or without that generator, my personal feeling is that test suites are still a better approach for the repository, since we're unlikely to start hooking a debugger into our CI. If you need some bad test case for debugging, you can keep a local copy of it, I have no intention of clobbering test/everything. If **we** need a particularly test case that often fails, we can keep a temporary copy in test/everything, but without adding it to TESTS or dist. Those tests are – as per the needs you ascribe to them – for debugging/development purposes, and people who need them should hence be comfortable using git to obtain them. We also shouldn't have any copypasta problems that come from the same test existing in two different locations, since those tests are static by design.

**Comment by LordYuuma** [comment]
> 1. One result per file. This is exactly what I sought to solve with my custom TAP bindings. However, that does not mean we can't write a test-driver, that has this based on SRFI-64, only that this specific driver likely is not what we're looking for. (Although I'm pretty sure, that `test-on-test-end-gnu` should report one result per test instead of per file, perhaps that's a bug?)

After running some tests locally, I've found this to not be the case. We can very well have multiple tests per file even with the existing driver.
> 2. Missing "ERROR". Our tests are normally wrapped in a `hard-error-if-exception` and also produce an ERROR if we abort or segfault on C. Such conditions can – at least according to this driver – not be replicated by SRFI-64. This is not too big of a problem though, especially if we can accept exceptions as soft errors. Hard errors from C will independantly be caught by the test environment.

We can live with that.
> 3. How to environment. Since the driver loads the script, we'd have to set up the environment around the driver, not the script. I'm pretty sure, guile-gcrypt also has the same issue and we can just copy their solution to it, though.

This is indeed tricky, but I have a solution, that works. Hence, I will rewrite these tests in terms of SRFI-64 and push them to another branch. This PR is hereby obsolete.
